stub: measure kernel/initrd/parameters into clean PCRs 11/12/13, and add "systemd-measure" tool to be able to pre-calculate values

poettering opened the pull request:
Currently PCR measurements on Linux are hard to pre-calculate. Let's do something about this: in sd-stub, let's cleanly measure the kernel/initrd/parameters we are about to invoke into clean, separate PCRs. That way, given any unified kernel image the expected value for PCR 12 
seen after boot can cleanly determined.

Then, let's add a tool to pre-calculate them, that takes similar parameters as the "objcopy" call that merges the blobs with the stub.

This is preparation to later being able to sign expect PCR values ahead of time, to allow cryptsetup TPM policies that bind to signatures of expected PCR values instead of the PCR values themselves, alleviating the PCR "brittleness" problem.

various other fixes, too.

poettering commented:
@Foxboron, @grawity This might be interesting to you?

(i.e. the systemd-measure tool is distantly related to https://github.com/grawity/tpm_futurepcr)

poettering commented:
/cc @medhefgo

Foxboron commented:
Ah, this is lovely!

It would be nicer if we could also implement self selected PCR values in `systemd-cryptsetup` or in the future have some way to integrate between them. But this can come later. See https://github.com/systemd/systemd/issues/19204

My only concern is that the Authenticode checksum of the PE binary is tedious to get right. With signatures we need to ensure it's padded to 8 bytes, but without a signature it should not be padded. Should test this with a signed kernel and an unsiged kernel just to be sure.

poettering commented:
> It would be nicer if we could also implement self selected PCR values in `systemd-cryptsetup` or in the future have some way to integrate between them. But this can come later. See #19204

Would be happy to match a patch for that. maybe we should extend the syntax for PCR policies like this:

0+4:sha1=123456789abcdef…+4:sha256=123456…+9@signed-pcr-hash.p7+11=/some/file.bin

i.e. this would encode:

* PCR 0 should be bound to hash values taken from the running system (i.e. the current syntax)
* PCR 4 should be bound to hash values, with two for sha1 and sha256 banks specified literally (i.e. syntax is `<nr>:<bank>=<value`)
* PCR 9 is defined through signature of PCR hashes (syntax is `<nr>@<signaturefile>`)
* PCR 11 is defined through blob. The specified is file is opened, and hashed for all relevant banks, and then the result used.

poettering commented:
(but anyway, that discussion should probably followed elsewhere)

bluca commented:
```
../src/boot/efi/stub.c: In function ‘efi_main’:
../src/boot/efi/stub.c:211:57: error: pointer targets in passing argument 1 of ‘strlena’ differ in signedness [-Werror=pointer-sign]
                               strlena(unified_sections[section]) + 1, /* including NUL byte */
      |                                         ~~~~~~~~~~~~~~~~^~~~~~~~~
      |                                                         |
      |                                                         const char *
In file included from ../src/boot/efi/stub.c:4:
/usr/include/efi/efilib.h:396:24: note: expected ‘const CHAR8 *’ {aka ‘const unsigned char *’} but argument is of type ‘const char *’
  396 |     IN CONST CHAR8    *s1
```

bluca commented on src/boot/efi/util.c:
```diff
@@ -743,3 +743,22 @@ EFI_STATUS make_file_device_path(EFI_HANDLE device, const char16_t *file, EFI_DE
+        static const char hex[16] = "0123456789abcdef";
+        _cleanup_free_ char16_t *buf = NULL;
+        const uint8_t *d = data;
+
```
assert on prefix and data

bluca commented on src/boot/bootctl.c:
```diff
@@ -613,22 +621,26 @@ static int status_variables(void) {
 
+        if (n_printed == 0) {
+                printf("No boot loaders listed in EFI Variables.\n\n");
+                return 0;
```
uneeded return

bluca reviewed (changes requested):
Superficial review, but looks great. How about a --json output switch for the measure tool, to facilitate further machine processing?

poettering commented on src/boot/efi/measure.c:
```diff
@@ -141,43 +141,79 @@ bool tpm_present(void) {
+        _cleanup_free_ char16_t *c = NULL;
+
+        if (description)
+                c = xstra_to_str(description);
```
in our efi code we have allocators that abort, given that we can't possibly recover from the inability to boot, given we are the only code running at that time, pretty much. The `x` in the name of `xstra_to_str()` is supposed to indicate that.

bluca replied:
ah, got it

medhefgo commented on src/boot/efi/stub.c:
```diff
@@ -204,6 +204,41 @@ EFI_STATUS efi_main(EFI_HANDLE image, EFI_SYSTEM_TABLE *sys_table) {
+                (void) tpm_log_event_ascii(
+                                TPM_PCR_INDEX_KERNEL_IMAGE,
+                                POINTER_TO_PHYSICAL_ADDRESS(sections[section]),
+                                strlena(sections[section]) + 1, /* including NUL byte */
```
```suggestion
                                strsize8(sections[section]),
```

medhefgo commented on src/fundamental/tpm-pcr.h:
```diff
@@ -0,0 +1,24 @@
+ * embedded initrd, and so on. In contrast to PCR 4 (which also contains this data, given the whole
+ * surrounding PE image is measured into it) this should be reasonably pre-calculatable, because it *only*
+ * consists of static data from the kernel PE image. */
+#define TPM_PCR_INDEX_KERNEL_IMAGE 11U
+
+/* This TPM PCR is where sd-stub extends the kernel command line and any passed credentials into. */
+#define TPM_PCR_INDEX_KERNEL_PARAMETERS 12U
+
+/* sd-stub used to write the kernel command line/credentials into PCR 8, in systemd <= 250. Let's provide for
+ * some compatibility. (Remove in 2023!) */
+#if EFI_TPM_PCR_COMPAT
+#define TPM_PCR_INDEX_KERNEL_PARAMETERS_COMPAT 8U
+#else
+#define TPM_PCR_INDEX_KERNEL_PARAMETERS_COMPAT UINT32_MAX
+#endif
+
+/* This TPM PCR is where we extend the initrd sysext images into which we pass to the booted kernel */
+#define TPM_PCR_INDEX_INITRD_SYSEXTS 13U
```
As an enum-enjoyer I have to ask: can we make these an enum? :D

medhefgo commented on src/fundamental/tpm-pcr.c:
```diff
@@ -0,0 +1,14 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "macro-fundamental.h"
```
What you need this for?

poettering replied:
it was an easy way to get the definition of NULL. will change to stddef.h

medhefgo commented on src/boot/efi/stub.c:
```diff
@@ -103,6 +103,13 @@ static EFI_STATUS combine_initrd(
+        static const uint64_t stub_features =
+                EFI_STUB_FEATURE_REPORT_BOOT_PARTITION |    /* We set LoaderDevicePartUUID */
+                EFI_STUB_FEATURE_PICK_UP_CREDENTIALS |      /* We pick up credentials from the boot partition */
+                EFI_STUB_FEATURE_PICK_UP_SYSEXTS |          /* We pick up system extensions from the boot partition */
```
If you add feature flags for these, the dtb feature is recent too, no?

poettering replied:
the feature flags are supposed to indicate stuff to userspace that it can make programmatical use of. i.e. it's more about reporting OS facing *API* features of the stub than just features of the stub that face elsewhere. I am not sure how an app running on the OS would take benefit of the dtb loading stuff, given the only way to get the dtb loaded is by linking it in via objcopy... it's a kernel builder facing feature, sure. But userspace API? i don't see it.

medhefgo commented on src/fundamental/efivars-fundamental.h:
```diff
@@ -13,6 +14,12 @@
 #define EFI_LOADER_FEATURE_LOAD_DRIVER             (UINT64_C(1) << 7)
 
+/* Features of the stub, i.e. systemd-stub */
+#define EFI_STUB_FEATURE_REPORT_BOOT_PARTITION     (UINT64_C(1) << 0)
+#define EFI_STUB_FEATURE_PICK_UP_CREDENTIALS       (UINT64_C(1) << 1)
+#define EFI_STUB_FEATURE_PICK_UP_SYSEXTS           (UINT64_C(1) << 2)
+#define EFI_STUB_FEATURE_THREE_PCRS                (UINT64_C(1) << 3)
```
`enum` feels left out :(

medhefgo commented on src/boot/efi/util.c:
```diff
@@ -743,3 +743,22 @@ EFI_STATUS make_file_device_path(EFI_HANDLE device, const char16_t *file, EFI_DE
+
+        buf[size*2] = 0;
+
+        log_error_stall(L"%s[%u]: %s", prefix, size, buf);
```
Needs `PRIuN` for size

behrmann commented on man/systemd-cryptenroll.xml:
```diff
@@ -296,11 +296,21 @@
 
+              <row>
+                <entry>11</entry>
+                <entry><citerefentry><refentrytitle>systemd-stub</refentrytitle><manvolnum>7</manvolnum></citerefentry> measures the ELF kernel image, embedded initrd and other payload of the PE image it is placed in in this PCR. Unlike PCR 4 (where the same data should be measured into), this PCR value should be easy to pre-calculate, as this only contains static parts of the PE binary. Use this ...
```
```suggestion
                <entry><citerefentry><refentrytitle>systemd-stub</refentrytitle><manvolnum>7</manvolnum></citerefentry> measures the ELF kernel image, embedded initrd and other payload of the PE image it is placed in into this PCR. Unlike PCR 4 (where the same data should be measured into), this PCR value should be easy to pre-calculate, as this only contains static parts of the PE binary. Use this PCR to bind TPM policies to a specific kernel image, possibly with an embedded initial RAM disk (initrd).</entry>
```

behrmann commented on man/systemd-measure.xml:
```diff
@@ -0,0 +1,154 @@
+  <refsect1>
+    <title>Commands</title>
+
+    <para>The follow commands are understood:</para>
```
```suggestion
    <para>The following commands are understood:</para>
```

behrmann commented on man/systemd-measure.xml:
```diff
@@ -0,0 +1,154 @@
+    <title>Examples</title>
+
+    <example>
+      <title>Generate a unified kernel image, and calculating the expected TPM PCR 11 value</title>
```
```suggestion
      <title>Generate a unified kernel image, and calculate the expected TPM PCR 11 value</title>
```
or
```suggestion
      <title>Generating a unified kernel image, and calculating the expected TPM PCR 11 value</title>
```

behrmann commented on man/systemd-stub.xml:
```diff
@@ -133,8 +140,8 @@
     <para>Also note that the Linux kernel will measure all initrds it receives into TPM PCR 9. This means
-    every type of initrd will be measured twice: the initrd embedded in the kernel image will be measured to
-    both PCR 4 and PCR 9; the initrd synthesized from credentials will be measured to both PCR 12 and PCR 9;
+    every type of initrd will be measured twice or thrice: the initrd embedded in the kernel image will be measured to
```
While I am a great fan of thrice, it is rather uncommon in most English vernaculars, at least if the intention is not to sound like some ancient tome of forbidden knowledge. :)
```suggestion
    every type of initrd will be measured twice or three times: the initrd embedded in the kernel image will be measured to
```
or
```suggestion
    every type of initrd will be measured two or three times: the initrd embedded in the kernel image will be measured to
```

bluca commented:
Some tests would also be good

poettering commented on src/boot/efi/measure.c:
```diff
@@ -141,43 +141,79 @@ bool tpm_present(void) {
 
-        /* PCR disabled */
-        if (pcrindex == UINT32_MAX)
+        /* If EFI_SUCCESS is returned, will initialize ret_measured to true if we actually measured
```
makes calling code simpler. i.e. the idea is that pcrindex == UINT32_MAX magically disables the measurement without this need special handling in all callers...

i.e. conceptually if we want something not to be treated as an error, we shouldn't return it as an error. because that forces calles to manually handle the errors manually even for clean codepaths, which is doable, but not ideal I think.

i.e. I am generally of the opinion that it's better to hide conditionalizations inside the functions to be conditionalized, instead of doing that in the caller. it's a scheme of "cleaner abstractions", as it allows you to see the general codeflow easier in the calling process, and not get lost in details about specific exceptional conditions...

Does that make any sense?

medhefgo replied:
I maintain that this makes all the code being touched ugly. Doing a `IN_SET(measure(...), EFI_SUCCESS, EFI_NOT_FOUND)` is imho much nicer to read than the `bool` dance you're doing and works similar to how `open()` works in userspace.

You also don't have to use `EFI_NOT_FOUND`. You can define a custom warning code and just use that. `EFI_STATUS` explicitly makes anything that doesn't have the MSB set a warning, which fits perfectly for this. In that case using `EFI_ERROR` to only filter for errors would even makes sense and keep the consumer side clean.

You should at least make the ret bool param mandatory where it makes sense (`pack_cpio` for example) to keep things simpler.

poettering replied:
Also note that there's stuff such as pack_cpio() that will set a `ret_measured`. Here in particular it becomes clear why the fact whether we actually measured something should not be propagated as an error code — it is not an *error* after all — but as a side-parameter. After all, the cpio packing should not abort just because the PCR measurement doesn't work.

Thats conceptually at the core of the problem: failure generally implies abortion of the regular execution paths of functions. And bypassing the PCR measurements should not do that. It should have almost no effect..

poettering commented on src/boot/efi/measure.c:
```diff
@@ -141,43 +141,79 @@ bool tpm_present(void) {
 }
 
-EFI_STATUS tpm_log_load_options(const char16_t *load_options) {
+EFI_STATUS tpm_log_event_ascii(uint32_t pcrindex, EFI_PHYSICAL_ADDRESS buffer, UINTN buffer_size, const char *description, bool *ret_measured) {
```
well, the efi tpm event log can be browsed with various tools, and they generally assume it contains a regular uefi string, i.e. utf16... the main reason for the decription is to make debugging easier, it is afte rall not part of the data that is measured. Hence it's helpful to stick to the format most events use, and that appears to be UTF-16...

poettering commented:
> Some tests would also be good

You are absolutely right, but this is a hard one, because we currently have no testsuite for sd-boot/sd-stub at all. We should though, but I am not sure this is the right PR to add that.

So, I am all for adding that, but let's maybe not block this PR for that.

bluca commented:
> > Some tests would also be good
> 
> You are absolutely right, but this is a hard one, because we currently have no testsuite for sd-boot/sd-stub at all. We should though, but I am not sure this is the right PR to add that.
> 
> So, I am all for adding that, but let's maybe not block this PR for that.

Maybe start with the `systemd-measure calculate` output? Some simple checks that say with an os-release and cmdline it spits out the expected values

poettering commented:
Force pushed a new version. Addressed the points raised. But not the enum-ification. I am not against converting the flags to an enum, but given the boot loader flags (which the stub flags take inspiration from) are macros right now, sounds like a clean-up to do separately, and then convert the whole thing.

Also, no test yet.

poettering commented:
Added test now, too.

bluca commented:
```
../src/src/boot/measure.c: In function ‘measure_pcr’:
../src/src/boot/measure.c:262:30: error: implicit declaration of function ‘read’; did you mean ‘readv’? [-Werror=implicit-function-declaration]
  262 |                         sz = read(fd, buffer, BUFFER_SIZE);
      |                              ^~~~
      |                              readv
../src/src/boot/measure.c:262:30: warning: nested extern declaration of ‘read’ [-Wnested-externs]
../src/src/boot/measure.c: In function ‘validate_stub’:
../src/src/boot/measure.c:429:21: error: implicit declaration of function ‘access’; did you mean ‘accept’? [-Werror=implicit-function-declaration]
  429 |                 if (access(p, F_OK) < 0) {
      |                     ^~~~~~
      |                     accept
../src/src/boot/measure.c:429:21: warning: nested extern declaration of ‘access’ [-Wnested-externs]
```

poettering commented:
Force pushed a new version. PTAL.

poettering commented:
(only fixes CI issues)

bluca commented:
> Superficial review, but looks great. How about a --json output switch for the measure tool, to facilitate further machine processing?

What about this?

poettering commented:
> > Superficial review, but looks great. How about a --json output switch for the measure tool, to facilitate further machine processing?
> 
> What about this?

Not against this. But maybe let's add this later. At this point, the whole tool is marked experimental in the docs, because I am not sure about the precise output. 

In particular I intend to add signing support, which makes the JSON stuff particularly interesting, since we need to output more complex structured data then. But I'd rather figure that out then, then doing half-assed JSON now that might need to be reworked later.

JSON means setting things in stone as an API, the current simply text based output much less so. Hence, I think JSON is what we should add once the main feature set we intend to provide here is in place and we can oversee what the API is going to be like.

poettering commented:
the CI failures appear unrelated.

medhefgo commented on src/boot/efi/measure.c:
```diff
@@ -141,43 +142,79 @@ bool tpm_present(void) {
-        EFI_TCG *tpm1;
+EFI_STATUS tpm_log_event(uint32_t pcrindex, EFI_PHYSICAL_ADDRESS buffer, UINTN buffer_size, const char16_t *description, bool *ret_measured) {
         EFI_TCG2 *tpm2;
+        EFI_STATUS ret;
```
Please rename to `err` for consistency everywhere.

medhefgo commented on src/boot/efi/util.c:
```diff
@@ -743,3 +743,25 @@ EFI_STATUS make_file_device_path(EFI_HANDLE device, const char16_t *file, EFI_DE
+        assert(prefix);
+        assert(data || size == 0);
+
+        /* Debugging helper — please keep this around, even if not used */
```
This since is for debugging, this should be behind `#ifdef EFI_DEBUG` and maybe be defined close to `debug_break`.

medhefgo commented on src/boot/efi/cpio.c:
```diff
@@ -432,23 +434,40 @@ EFI_STATUS pack_cpio(
         for (UINTN i = 0; i < n_tpm_pcr; i++) {
+                bool m;
+
+                if (tpm_pcr[i] == UINT32_MAX) /* Disabled */
```
This is already done by tpm_log_event, why replicate?

poettering replied:
if we measure multiple things to a PCR and one measurement isn't done we want to treat that as incompletely measured, i.e. set ret_measured = false. But if we measure the same thing to multiple PCRs, and one of that isn't done because explicitly disabled, then we still want to consider the measurement as having taken place.

i.e. measurement to multiple PCRs is redundant, and it's fine if one of them is disabled.

measurement of multiple blocks where one is skipped is *not* redundant though, and we should handle that differently.

poettering commented:
> You should at least make the ret bool param mandatory where it makes sense (pack_cpio for example) to keep things simpler.

it shouldn't be mandatory though. I intend to soon extend the whole logic so that we can embed signed PCR hash values in the PE image, and those should not be measured (because that would be cycle then), but still be packed up in a cpio.

(i.e. my goal here is that now that we have PCR hashes that can be pre-calculate people can use them to implement signature based PCR hash schemes, so that kernel images can be updated easily without this requiring re-enrollment of the hashes. These signatures should be included in the PE image, but unlike the rest of the sections which should be measured into PCRs, this signature section shouldn't be, because it already contains the expected results of the measurements, and making it input for the measurements is hence a bit useless)

Or in other words unified kernels would be built like this:

1. We prepare the components: kernel, initrd, boot splash, dtb, …
2. We pre-calculate the PCR hashes for these components
3. We sign the PCR hashes with a priv/pub key pair of our choosing
4. We now build the PE image from stub, from the mentioned components, and the signed expected hash values.
5. We sign the whole thing with SecureBoot certificates

On boot, the hash/sig data is taken by sd-stub, tunred into an initrd, and passed to the booted kernel. There it can be consumed by cryptsetup for example, to unlock a disk.

Net result: you can easily set up encrypted disks that can only be decrypted if a very specific series of kernels are booted. This is nice since SecureBoot certificates are typically excessively wide in focus (i.e, anyone who asks nicely will get a kernel/boot loader signed by MSFT), but with this we can tightly control with a very narrow focus which kernels actually get access to a disk.

And that's excellent for "opportunistic encryption", i.e. systems which are auto-encrypted by default, and whose keys are locked to TPM only and thus can be unlocked non-interactively at boot.

poettering commented:
Force pushed new version. Addressed some of @medhefgo's points, but not the `ret_measure` thing, nor the `enum` thing. (I disagree with the former, and the latter sounds good to me, but in a later commit, since it should probably converted together with other stuff not directly related to this PR)

poettering commented:
CI failures appear unrelated...
